base/linalg/generic.jl

```diff
@@ -431,16 +431,15 @@ julia> vecnorm([1 2 3 4 5 6 7 8 9])
 ```
 """
 function vecnorm(itr, p::Real=2)
-    isempty(itr) && return float(real(zero(eltype(itr))))
+    isempty(itr) && return float(norm(zero(eltype(itr))))
```

Comment:
Notice that the two type computations for the empty case and the `p=0` case are slightly different. The empty case uses `zero(Type)` whereas the `p=0` case avoids `zero(Type)` which allows us to compute
```julia
julia> norm(Float64[])
0.0
```
as well as
```julia
julia> norm([[1,2], [3,4]], 0)
2.0
```
I think we'd have to give up on one of them if we want to use return type annotation instead.